src/OmniSharp.MiscellanousFiles/MiscellanousFiles.cs

```diff
@@ -0,0 +1,115 @@
+        IEnumerable<string> IProjectSystem.Extensions => new[] { miscFileExtension };
+        public bool EnabledByDefault { get; } = true;
+
+        private readonly ConcurrentDictionary<string, ProjectInfo> _projects = new ConcurrentDictionary<string, ProjectInfo>();
+        private readonly OmniSharpWorkspace _workspace;
+        private readonly IFileSystemWatcher _fileSystemWatcher;
+        private readonly FileSystemHelper _fileSystemHelper;
+        private readonly ProjectSystem _projectSystem;
+        private readonly ILogger _logger;
+
+        [ImportingConstructor]
+        public MiscellanousFiles(OmniSharpWorkspace workspace, IFileSystemWatcher fileSystemWatcher, FileSystemHelper fileSystemHelper,
+            ILoggerFactory loggerFactory, [ImportMany] IEnumerable<Lazy<IProjectSystem, ProjectSystemMetadata>> projectSystems)
+        {
+            _workspace = workspace;
+            _fileSystemWatcher = fileSystemWatcher ?? throw new ArgumentNullException(nameof(fileSystemWatcher));
+            _fileSystemHelper = fileSystemHelper;
+            _logger = loggerFactory.CreateLogger<MiscellanousFiles>();
+            _projectSystem = (ProjectSystem)projectSystems.Where(ps => ps.Metadata.Name == nameof(ProjectSystem)).First().Value;
+        }
+
+        Task<object> IProjectSystem.GetProjectModelAsync(string filePath)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        Task<object> IProjectSystem.GetWorkspaceModelAsync(WorkspaceInformationRequest request)
+        {
+            return Task.FromResult<object>(null);
+        }
+
+        void IProjectSystem.Initalize(IConfiguration configuration)
+        {
+            var allFiles = _fileSystemHelper.GetFiles("**/*.cs");
+            foreach (var filePath in allFiles)
+                AddIfMiscellanousFile(filePath);
+
+            _fileSystemWatcher.Watch(miscFileExtension, OnMiscellanousFileChanged);
+        }
+
+        private async void AddIfMiscellanousFile(string filePath)
+        {
+            //wait for the project system to get initialised
+            await _projectSystem.HasCompletedUpdateRequest();
+            if (_workspace.GetDocument(filePath) == null)
+            {
+                string assemblyName = Guid.NewGuid().ToString("N");
+                var projectInfo = ProjectInfo.Create(
+                   filePath: filePath,
+                   id: ProjectId.CreateNewId(),
+                   version: VersionStamp.Create(),
+                   name: Path.GetFileName(filePath),
+                   metadataReferences: new MetadataReference[] { MetadataReference.CreateFromFile((typeof(object).Assembly).Location) },
+                   assemblyName: assemblyName,
+                   language: Language);
+
+                _workspace.AddProject(projectInfo);
+                _workspace.AddMiscellanousFileDocument(projectInfo.Id, filePath);
+                _projects[filePath] = projectInfo;
```

Comment:
It strikes me as overkill to add a project for every single miscellaneous file. In Visual Studio, I believe there's a single miscellaneous project that all misc files are added to. @jasonmalinowski would know best.

Reply:
In Visual Studio we do create a separate project per individual file. This was largely to replicate the previous C# language service that did effectively the same in my opinion. I've long thought about moving them into one project though, so opportunistically things like go to def will move between files if you happen to have multiple files from the same project open.